### ISSUE TITLE
Suggestion: Edit Mode

### ISSUE THREAD

**Issue by svanderbleek** [philc/vimium#1853]
Like Visual mode works with Vimium now it would be great to have an edit mode for when you are working in a textarea/contenteditable or whenever really, I use vi bindings with terminal readline. So for example `E` could go to edit mode where navigation is Vim text style and `esc` would go back to browser navigation mode.

**Comment by smblott-github** [comment]
@svanderbleek.  The code base does actually contain an edit mode, although it's wholly disabled in the Chrome Store version.  It's mentioned here: #1483.

The current implementation is fairly basic.  It works on GMail, which is good, but it's fairly limited in terms of the vim-like bindings supported.

I've been wondering recently what to do with this.  If we enable it in the CS build, then we have to maintain it -- and we have to bring the UX up to scratch.  The thing is ... I wrote this myself, but don't actually use it myself!  (I use [this](https://chrome.google.com/webstore/detail/text-aid-too/klbcooigafjpbiahdjccmajnaehomajc)).  That tells you something in itself.

I think our options are:

1. Strip edit mode from the code base.  There are other/better ways to achieve the same thing.
2. Enable the current edit mode (and maintain it).
3. Embed one of the other (very-good) JavaScript vim emulators. (Embedding other projects' code is not something we've done before.)

To me, 1 and 3 seem like the best directions to go.

**Comment by smblott-github** [comment]
Actually, on brief reflection, maybe we **should**:

- enable edit mode in master
- try it out for a month or two and work on the UX
- publish it as an experimental feature
- ... and see what happens.

If we could get the UX right, it would make a pretty cool signature feature for a release.

(@mrmr1993, @philc, ... What do you think?)

**Comment by mrmr1993** [comment]
I have a major rewrite of visual mode/edit mode in the works, which should make the code simpler, more maintainable, and more easily extensible.

Since edit mode doesn't need to do much more than visual mode, I'd think that once this is done, the maintainance burden of edit mode will be nothing compared to visual mode.

I can submit PRs as I go, or one big one when I feel done, if this is the way we want to go. What do you think?

**Comment by smblott-github** [comment]
Perhaps float your ideas either here or in a new issue to get feedback before progressing too far, @mrmr1993.

**Comment by mrmr1993** [comment]
The ideas are quite simple:

* move all the selection-related functions into a helper class
* simplify them down, make them general purpose
* create subclasses, one for `<input>`s/`<textarea>`s, one for anything else, so that we can use different codepaths easily
* create `moveSelectionForwardToRegexp` and `moveSelectionBackwardToRegexp`, using it to rewrite all the moment commands much more simply
* fix the confusing clipboard logic
* split visual mode into two linked modes, one movement and one actions, since there's not much overlap in implementation
* (just for good measure) use #1775 to remove all the dedicated keyhandling logic

There's a few other things that we could do, but I think these are most important and straightforward for maintainability.

**Comment by smblott-github** [comment]
> move all the selection-related functions into a helper class

That's kind of the role of the `Movement` class, is it not?

> simplify them down, make them general purpose

I kind of thought `Movement` was general purpose.  The same movements are used by visual, edit and caret modes.

Perhaps `Movement` could be refactored a bit to simplify it.

> create subclasses, one for `<input>s/<textarea>s`, one for anything else, so that we can use different codepaths easily

I don't understand this.  Could you re-state it?

> create moveSelectionForwardToRegexp and moveSelectionBackwardToRegexp, using it to rewrite all the moment commands much more simply

That sounds like a good idea.  The code for some movements (e.g., word movements) is pretty horrible currently.

> fix the confusing clipboard logic

I was going to post this separately.

In fact, I think this might have fixed itself.  In Linux, it used to be the case that simply selecting text in Chrome copied it to the clipboard.  So, in visual mode, `w`, `w`, `w` caused the clipboard to be updated three times.  The clipboard logic is all about preventing (or, rather, reverting) these updates.

In recent versions of Chrome, I'm no longer seeing the original behaviour.  You have to use `Ctrl-c` explicitly to copy.  So it might be that we can just ditch the clipboard code.

> split visual mode into two linked modes, one movement and one actions, since there's not much overlap in implementation

Maybe.  Some the action code in `Movement` is made particularly complicated because its behaviour is tweaked depending upon the parent mode (visual, edit, etc).  This can probably be simplified.

> (just for good measure) use #1775 to remove all the dedicated keyhandling logic

There has been surprisingly little call for remapping of visual-mode keys.  There could be several reasons for that:

- nobody's using visual mode
- since visual mode only implements the most basic vim movements, everybody's familiar with them anyway, or
- again, since we only have basic movements, these are movements which few people rebind anyway.

**Comment by svanderbleek** [comment]
I use visual mode for copying text, but it's kind of useless right now except for line visual mode since it's missing `f` and `t` etc it seems.

**Comment by mrmr1993** [comment]
> > move all the selection-related functions into a helper class

> That's kind of the role of the `Movement` class, is it not?

That, and managing the commands. I feel like muddling them together is to mix high-level logic and low-level implementation. Plus, this allows the subclasses I was talking about, which will help readability massively IMO.

> I kind of thought `Movement` was general purpose. 

Trying to add a new command (eg. `a(`) seems to require code duplication and a whole group of changes, rather than a simple description. I'd say it's more made-for-purpose than general purpose.

> > create subclasses, one for `<input>`s/`<textarea>`s, one for anything else, so that we can use different codepaths easily

> I don't understand this. Could you re-state it?

At the moment, some movements are done by stepping character by character and checking the character added/removed from the selection. Instead, it makes more sense to me to
* use `selectionStart`, `selectionEnd` and `selectionDirection` to change the selection and `value` to work out where to place the cursor for `<input>`s/`<textarea>`s, and
* use `Selection.collapse` and `Selection.extend` to change the selection and `Range.toString()` to work out where to place the cursor for all other selection contexts.

A great example of where this would help is reversing the selection, and it would also get us out of the habit of changing the selection in order to read a value (the next character), often just to change it back again.

> The clipboard logic is all about preventing (or, rather, reverting) these updates.

Maybe I've missed something here, but I can't think why we wouldn't just call a `copy` function when we want to copy. The current system of 'always copy the selection to the clipboard but stop it from working sometimes' didn't make a lot of sense to me.

> There has been surprisingly little call for remapping of visual-mode keys.

That's not what #1775 does. But it does allow us to remove all the dedicated key handling logic.

**Comment by rrsahu** [comment]
For example, if you wanted to shift car then we have car providers. You don’t have to panic on price range as our solutions are very reasonably priced and given up to the mark.

<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers in Bangalore">Packers and Movers in Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers in Pune">Packers and Movers in Pune</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers in Mumbai">Packers and Movers in Mumbai</a>

**Comment by rrsahu** [comment]
Secondly, you don’t have to take additional break from your work hours to package products and moving. In this way, we help you in avoiding perform reduction too. 

<a href="http://www.expert5th.in/packers-and-movers-chennai/" title="Packers and Movers Chennai">Packers and Movers Chennai</a>
<a href="http://www.expert5th.in/packers-and-movers-noida/" title="Packers and Movers Noida">Packers and Movers Noida</a>
<a href="http://www.expert5th.in/packers-and-movers-navimumbai/" title="Packers and Movers Navi Mumbai">Packers and Movers Navi Mumbai</a>
<a href="http://www.expert5th.in/packers-and-movers-thane/" title="Packers and Movers Thane">Packers and Movers Thane</a>
<a href="http://www.expert5th.in/packers-and-movers-ghaziabad/" title="Packers and Movers Ghaziabad">Packers and Movers Ghaziabad</a>
<a href="http://www.expert5th.in/packers-and-movers-faridabad/" title="Packers and Movers Faridabad">Packers and Movers Faridabad</a>

**Comment by rrsahu** [comment]
You just focus on other effective perform and keep the rest on us. Working with well-known packers movers is not only practical but extremely safe option too. We are providing reliable protection to our customers.

<a href="http://www.expert5th.in/packers-and-movers-gurgaon/" title="Packers and Movers Gurgaon">Packers and Movers Gurgaon</a>
<a href="http://www.expert5th.in/packers-and-movers-delhi/" title="Packers and Movers Delhi">Packers and Movers Delhi</a>
<a href="http://www.expert5th.in/packers-and-movers-pune/" title="Packers and Movers Pune">Packers and Movers Pune</a>

**Comment by rrsahu** [comment]
We are best support agency in Bangalore providing best offers for moving. Packages can be customized depending on your moving specifications. 

<a href="http://www.expert5th.in/packers-and-movers-hyderabad/" title="Packers and Movers Hyderabad">Packers and Movers Hyderabad</a>
<a href="http://www.expert5th.in/packers-and-movers-bangalore/" title="Packers and Movers Bangalore">Packers and Movers Bangalore</a>
<a href="http://www.expert5th.in/packers-and-movers-mumbai/" title="Packers and Movers Mumbai">Packers and Movers Mumbai</a>

**Comment by mcp292** [comment]
I have been seeking such a feature! Especially given that [older solutions no longer work](https://stackoverflow.com/a/75665/4682839).

I'm on Firefox on Linux.

**Comment by mcp292** [comment]
It would also be amazing to add register support so you could `"2p` in the browser text box, or `3"p` what you yanked in the browser, back in Vim.

**Comment by mcp292** [comment]
https://github.com/philc/vimium/issues/1059
